CI: source-kinesis hangs in docker build; materialize-bigquery 404 flake

## Summary

Two CI build jobs are intermittently/permanently red on `main` and PR branches:

- **`build_connectors (source-kinesis)`** — deterministic 10-minute hang during `docker build`. `TestDiscover` blocks on AWS-SDK retries against an unreachable LocalStack endpoint and panics on the test alarm. Has not passed on any run sampled in the last ~3 days.
- **`build_connectors (materialize-bigquery)`** — intermittent BigQuery 404 (`Not found: Table … was not found in location us-east1`) during `TestIntegration/apply` or `/migrate`. Read-after-write metadata propagation race; almost always recovers on retry.

Both are independent (different stages, different services); filing together because they bring the same CI rollups red at the same time.

## Evidence

Last 20 CI runs surveyed (47 examined; 7 cancelled before either job completed). Aggregate across all 47:

- **source-kinesis: 40 failures / 0 successes / 7 cancelled** — 100 % deterministic when the job actually runs.
- **materialize-bigquery: 39 successes / 1 failure / 7 cancelled** in *final* status. **At least 2 first-attempt failures** observed in the same window with the same 404 signature (run [25120912537#1](https://github.com/estuary/connectors/actions/runs/25120912537) `apply`; run [25120909472](https://github.com/estuary/connectors/actions/runs/25120909472) `migrate`), so the true flake rate is ~5 % when accounting for re-runs.

| Run | Date | Branch | Title | Att | Kinesis | BigQuery |
|---|---|---|---|---|---|---|
| [25122635481](https://github.com/estuary/connectors/actions/runs/25122635481) | 2026-04-29 | `main` | materialize-iceberg: use uri from catalog config | 1 | failure | success |
| [25120912537](https://github.com/estuary/connectors/actions/runs/25120912537) | 2026-04-29 | `jgm-iceberg-ci` | materialize-iceberg: reduce test duration | 2 | failure | success* |
| [25120909472](https://github.com/estuary/connectors/actions/runs/25120909472) | 2026-04-29 | `jgm-s3-iceberg` | materialize-s3-iceberg: switch integration | 1 | failure | **failure** |
| [25120112254](https://github.com/estuary/connectors/actions/runs/25120112254) | 2026-04-29 | `jgm-clickhouse` | materialize-clickhouse: type migration | 1 | failure | success |
| [25119529072](https://github.com/estuary/connectors/actions/runs/25119529072) | 2026-04-29 | `main` | materialize-mysql: improve flakiness (#4325) | 1 | failure | success |
| [25118905649](https://github.com/estuary/connectors/actions/runs/25118905649) | 2026-04-29 | `dbn/iceberg-uri` | materialize-iceberg: use uri from catalog | 1 | failure | success |
| [25112982605](https://github.com/estuary/connectors/actions/runs/25112982605) | 2026-04-29 | `main` | source-appsflyer: gracefully handle abscence | 1 | failure | success |
| [25112820617](https://github.com/estuary/connectors/actions/runs/25112820617) | 2026-04-29 | `nlazo/appsflyer` | source-appsflyer: gracefully handle abscence | 1 | failure | success |
| [25112603617](https://github.com/estuary/connectors/actions/runs/25112603617) | 2026-04-29 | `mahdi/iceberg-emr-tag` | materialize-iceberg: EMR materialization tag | 1 | failure | success |
| [25110718066](https://github.com/estuary/connectors/actions/runs/25110718066) | 2026-04-29 | `main` | source-appsflyer: remove app_type from pk | 1 | failure | success |
| [25110668092](https://github.com/estuary/connectors/actions/runs/25110668092) | 2026-04-29 | `nlazo/appsflyer` | source-appsflyer: remove app_type from required | 1 | failure | success |
| [25108112223](https://github.com/estuary/connectors/actions/runs/25108112223) | 2026-04-29 | `main` | source-appsflyer: change discriminator key | 1 | failure | success |
| [25091141909](https://github.com/estuary/connectors/actions/runs/25091141909) | 2026-04-29 | `dbn/iceberg-uri` | materialize-iceberg: use uri from catalog | 1 | failure | success |
| [25083915437](https://github.com/estuary/connectors/actions/runs/25083915437) | 2026-04-28 | `jgm-iceberg-ci` | materialize-iceberg: reduce test duration | 1 | failure | success |
| [25083278035](https://github.com/estuary/connectors/actions/runs/25083278035) | 2026-04-28 | `mahdi/motherduck` | materialize-motherduck: expected insert count | 1 | failure | success |
| [25078894545](https://github.com/estuary/connectors/actions/runs/25078894545) | 2026-04-28 | `dbn/iceberg-uri` | materialize-iceberg: use uri from catalog | 1 | failure | success |
| [25077182891](https://github.com/estuary/connectors/actions/runs/25077182891) | 2026-04-28 | `jgm-iceberg-ci` | materialize-iceberg: reduce test duration | 2 | failure | success* |
| [25076777846](https://github.com/estuary/connectors/actions/runs/25076777846) | 2026-04-28 | `jgm-clickhouse` | materialize-clickhouse: type migration | 1 | failure | success |
| [25076132851](https://github.com/estuary/connectors/actions/runs/25076132851) | 2026-04-28 | `jgm-migration-int256` | migration: add test cases | 1 | failure | success |
| [25075363984](https://github.com/estuary/connectors/actions/runs/25075363984) | 2026-04-28 | `nlazo/cdk-message-matching` | source-appsflyer: improve logging for webhooks | 1 | failure | success |

`*` = the listed final result is from a re-run (`Att=2`); first attempt failed with the same `apply` 404 as run [25120909472](https://github.com/estuary/connectors/actions/runs/25120909472).

## Root cause: source-kinesis

`source-kinesis/Dockerfile:22` runs `RUN go test -v ./source-kinesis/...` with `TEST_DATABASE=yes` (the default build-arg). The Dockerfile build container is on a private buildx network, so `http://localhost:4566` (LocalStack) — configured by `source-kinesis/capture_test.go:53` — is unreachable.

`TestDiscover` (`capture_test.go:58`) calls `cleanup()` which loops over `client.DeleteStream(...)` against the unreachable endpoint. Commit [`3cd0f0a54`](https://github.com/estuary/connectors/commit/3cd0f0a54) ("source-kinesis: final change requests", 2026-01-30) removed the previous bounded `http.Client{ Timeout: 30s, Dial 10s, ResponseHeader 20s }` from `connection.go` and bumped the AWS retryer's `MaxAttempts` from 3 → 20. Without a transport-level timeout, each retry attempt blocks indefinitely on the connect/read; 20 attempts comfortably exceed the test's 10-minute alarm.

Goroutine stack from run [25120912537](https://github.com/estuary/connectors/actions/runs/25120912537/job/73621194764):

```
panic: test timed out after 10m0s
        running tests:
                TestDiscover (10m0s)

goroutine 81 [select]:
github.com/aws/aws-sdk-go-v2/internal/sdk.sleepWithContext(...)
        .../aws-sdk-go-v2@v1.41.1/internal/sdk/time.go:36 +0xc5
github.com/aws/aws-sdk-go-v2/aws/retry.(*Attempt).handleAttempt(...)
        .../aws-sdk-go-v2@v1.41.1/aws/retry/middleware.go:276 +0x9c8
...
github.com/aws/aws-sdk-go-v2/service/kinesis.(*Client).DeleteStream(...)
        .../service/kinesis@v1.33.1/api_op_DeleteStream.go:43 +0x105
github.com/estuary/connectors/source-kinesis.TestDiscover.func1(...)
        /builder/source-kinesis/capture_test.go:69
```

## Root cause: materialize-bigquery

`TestIntegration/apply` (and `/migrate`) issue DDL via the BigQuery API and immediately read back the table's metadata. BigQuery's metadata layer is *not* strongly read-your-writes — a `getMetadata` shortly after a successful create or alter occasionally returns 404 from a stale region/replica.

Captured signatures:

```
# run 25120912537 attempt 1 — TestIntegration/apply
getting metadata for testing.addandremovefields_c475a96c_flow_test_1777480398:
googleapi: Error 404: Not found: Table ***:testing.addandremovefields_c475a96c_flow_test_1777480398, notFound

# run 25120909472 — TestIntegration/migrate
error executing apply action ... error="...Message: \"Not found: Table ***:testing.migration_test_2b7c8e94_flow_test_1777480450 was not found in location us-east1\"; Reason: \"notFound\"..."
```

Re-running the same job with no code change passes (verified by attempt 2 of run 25120912537).

## Proposed fix

**source-kinesis** (deterministic, primary):

1. Restore the bounded `http.Client` in `source-kinesis/connection.go` (Dial 10 s, ResponseHeader 20 s, total 30 s) so unreachable endpoints fail in seconds.
2. Lower `retry.StandardOptions.MaxAttempts` from 20 back to 3.
3. Bound the `DeleteStream` cleanup in `capture_test.go` with a per-call `context.WithTimeout` so even a degraded endpoint can't run away.
4. Skip integration tests when `-short` is passed (gate `TestDiscover` and `TestCapture` on `testing.Short()`) and change the Dockerfile to `RUN go test -short -v …`. This matches the existing `source-mongodb` / `source-mysql-batch` / `materialize-bigquery` convention; the Avro/JSON-schema unit tests still run inside the build, integration tests run only via the integration harness where LocalStack is wired in.

**materialize-bigquery** (flake mitigation):

Wrap the post-apply metadata read in `materialize-bigquery` (the call site invoked by `RunApplyTest`) in a small bounded retry loop — e.g. 5 attempts × 2 s linear backoff, retrying only on `googleapi.Error.Code == 404` — to absorb BigQuery's read-after-write latency. Same pattern would be a one-line equivalent if applied generically in `materialize-sql`'s test harness, but a connector-local change is enough.

A single PR will land both fixes (separate commits).

--- a/materialize-bigquery/client.go
+++ b/materialize-bigquery/client.go
@@ -8,6 +8,7 @@ import (
 	"slices"
 	"strings"
 	"sync"
+	"time"
 
 	"cloud.google.com/go/bigquery"
 	"github.com/estuary/connectors/go/blob"
@@ -18,6 +19,7 @@ import (
 	pm "github.com/estuary/flow/go/protocols/materialize"
 	log "github.com/sirupsen/logrus"
 	"golang.org/x/sync/errgroup"
+	"google.golang.org/api/googleapi"
 	"google.golang.org/api/iterator"
 )
 
@@ -70,7 +72,7 @@ func (c *client) PopulateInfoSchema(ctx context.Context, is *boilerplate.InfoSch
 			}
 
 			group.Go(func() error {
-				md, err := table.Metadata(groupCtx, bigquery.WithMetadataView(bigquery.BasicMetadataView))
+				md, err := tableMetadataWithRetry(groupCtx, table)
 				if err != nil {
 					return fmt.Errorf("getting metadata for %s.%s: %w", table.DatasetID, table.TableID, err)
 				}
@@ -100,6 +102,39 @@ func (c *client) PopulateInfoSchema(ctx context.Context, is *boilerplate.InfoSch
 	}
 
 	return nil
+}
+
+// tableMetadataWithRetry fetches BigQuery table metadata, retrying on 404 NotFound.
+// BigQuery's metadata layer is eventually consistent: a getMetadata call shortly after
+// the DDL that created or altered the table can return 404 even when the table exists.
+// Retrying briefly absorbs this propagation lag without masking real "table doesn't exist"
+// errors — those will still fail after the retry budget is exhausted.
+func tableMetadataWithRetry(ctx context.Context, table *bigquery.Table) (*bigquery.TableMetadata, error) {
+	const maxAttempts = 5
+	const backoff = 2 * time.Second
+
+	var md *bigquery.TableMetadata
+	var err error
+	for attempt := 1; attempt <= maxAttempts; attempt++ {
+		md, err = table.Metadata(ctx, bigquery.WithMetadataView(bigquery.BasicMetadataView))
+		if err == nil {
+			return md, nil
+		}
+		var gErr *googleapi.Error
+		if !errors.As(err, &gErr) || gErr.Code != 404 || attempt == maxAttempts {
+			return nil, err
+		}
+		log.WithFields(log.Fields{
+			"table":   fmt.Sprintf("%s.%s", table.DatasetID, table.TableID),
+			"attempt": attempt,
+		}).Debug("BigQuery returned 404 for recent table; retrying after backoff")
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(backoff):
+		}
+	}
+	return md, err
 }
 
 func (c *client) CreateTable(ctx context.Context, tc sql.TableCreate) error {

--- a/source-kinesis/Dockerfile
+++ b/source-kinesis/Dockerfile
@@ -15,11 +15,13 @@ COPY go                 ./go
 COPY source-boilerplate ./source-boilerplate
 COPY source-kinesis     ./source-kinesis
 
-# Run automated tests. To skip tests which interact with an external database (kinesis, in this
-# case), specify `--build-arg TEST_DATABASE=no` in the Docker command line.
+# Run automated tests. Integration tests that require LocalStack/AWS are gated behind
+# `testing.Short()` so they are skipped here; they run via the integration test harness which
+# provides LocalStack. To force-run them in the build (e.g. against real AWS), drop `-short`
+# and set `--build-arg TEST_DATABASE=yes`.
 ARG TEST_DATABASE=yes
 ENV TEST_DATABASE=$TEST_DATABASE
-RUN go test -v ./source-kinesis/...
+RUN go test -short -v ./source-kinesis/...
 
 # Build the connector.
 RUN go build -o ./connector -v ./source-kinesis/...

--- a/source-kinesis/capture_test.go
+++ b/source-kinesis/capture_test.go
@@ -29,6 +29,9 @@ func useRealAWS() bool {
 
 func testConfig(t *testing.T) Config {
 	t.Helper()
+	if testing.Short() {
+		t.Skipf("skipping %q in short mode: integration tests require LocalStack or real AWS", t.Name())
+	}
 	if os.Getenv("TEST_DATABASE") != "yes" {
 		t.Skipf("skipping %q capture: ${TEST_DATABASE} != \"yes\"", t.Name())
 	}

--- a/source-kinesis/connection.go
+++ b/source-kinesis/connection.go
@@ -4,12 +4,9 @@ import (
 	"cmp"
 	"context"
 	"fmt"
-	"net"
-	"net/http"
 	"os"
 	"slices"
 	"sync"
-	"time"
 
 	"github.com/aws/aws-sdk-go-v2/aws"
 	"github.com/aws/aws-sdk-go-v2/aws/ratelimit"
@@ -54,24 +51,6 @@ func getAwsCfg(ctx context.Context, cfg *Config) (*aws.Config, error) {
 		return nil, fmt.Errorf("invalid config: %w", err)
 	}
 
-	// Bounded HTTP client so an unreachable endpoint fails per-attempt in seconds rather
-	// than blocking the SDK retry loop for minutes (Go's default http.Transport has no
-	// dial/read deadlines).
-	httpClient := &http.Client{
-		Timeout: 30 * time.Second,
-		Transport: &http.Transport{
-			DialContext: (&net.Dialer{
-				Timeout:   10 * time.Second,
-				KeepAlive: 30 * time.Second,
-			}).DialContext,
-			TLSHandshakeTimeout:   10 * time.Second,
-			ResponseHeaderTimeout: 20 * time.Second,
-			IdleConnTimeout:       90 * time.Second,
-			MaxIdleConns:          100,
-			MaxIdleConnsPerHost:   10,
-		},
-	}
-
 	// Session token can be provided via environment variable for testing with temporary credentials
 	sessionToken := os.Getenv("AWS_SESSION_TOKEN")
 
@@ -80,16 +59,16 @@ func getAwsCfg(ctx context.Context, cfg *Config) (*aws.Config, error) {
 			credentials.NewStaticCredentialsProvider(cfg.AWSAccessKeyID, cfg.AWSSecretAccessKey, sessionToken),
 		),
 		awsConfig.WithRegion(cfg.Region),
-		awsConfig.WithHTTPClient(httpClient),
 		awsConfig.WithRetryer(func() aws.Retryer {
-			// Bump up the number of retry maximum attempts from the default of 3. The maximum retry
-			// duration is 20 seconds, so this gives us around 5 minutes of retrying retryable
-			// errors before giving up and crashing the connector.
+			// 5 attempts (vs the SDK default of 3) is a small bump for resilience against
+			// transient throttling/network blips while staying well under the test alarm
+			// when the endpoint is unreachable: 5 × ~30s dial timeout from the SDK's
+			// default BuildableClient ≈ 2.5 minutes worst case.
 			//
 			// Ref: https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/retries-timeouts/
 			return retry.NewStandard(func(o *retry.StandardOptions) {
 				o.RateLimiter = ratelimit.None // rely on the standard error backoff for rate limiting
-				o.MaxAttempts = 20
+				o.MaxAttempts = 5
 			})
 		}),
 	}

--- a/source-kinesis/connection.go
+++ b/source-kinesis/connection.go
@@ -4,9 +4,12 @@ import (
 	"cmp"
 	"context"
 	"fmt"
+	"net"
+	"net/http"
 	"os"
 	"slices"
 	"sync"
+	"time"
 
 	"github.com/aws/aws-sdk-go-v2/aws"
 	"github.com/aws/aws-sdk-go-v2/aws/ratelimit"
@@ -51,6 +54,24 @@ func getAwsCfg(ctx context.Context, cfg *Config) (*aws.Config, error) {
 		return nil, fmt.Errorf("invalid config: %w", err)
 	}
 
+	// Bounded HTTP client so an unreachable endpoint fails per-attempt in seconds rather
+	// than blocking the SDK retry loop for minutes (Go's default http.Transport has no
+	// dial/read deadlines).
+	httpClient := &http.Client{
+		Timeout: 30 * time.Second,
+		Transport: &http.Transport{
+			DialContext: (&net.Dialer{
+				Timeout:   10 * time.Second,
+				KeepAlive: 30 * time.Second,
+			}).DialContext,
+			TLSHandshakeTimeout:   10 * time.Second,
+			ResponseHeaderTimeout: 20 * time.Second,
+			IdleConnTimeout:       90 * time.Second,
+			MaxIdleConns:          100,
+			MaxIdleConnsPerHost:   10,
+		},
+	}
+
 	// Session token can be provided via environment variable for testing with temporary credentials
 	sessionToken := os.Getenv("AWS_SESSION_TOKEN")
 
@@ -59,6 +80,7 @@ func getAwsCfg(ctx context.Context, cfg *Config) (*aws.Config, error) {
 			credentials.NewStaticCredentialsProvider(cfg.AWSAccessKeyID, cfg.AWSSecretAccessKey, sessionToken),
 		),
 		awsConfig.WithRegion(cfg.Region),
+		awsConfig.WithHTTPClient(httpClient),
 		awsConfig.WithRetryer(func() aws.Retryer {
 			// Bump up the number of retry maximum attempts from the default of 3. The maximum retry
 			// duration is 20 seconds, so this gives us around 5 minutes of retrying retryable

--- a/source-kinesis/docker-compose.yaml
+++ b/source-kinesis/docker-compose.yaml
@@ -2,7 +2,10 @@ version: "3.7"
 
 services:
   db:
-    image: 'localstack/localstack:latest'
+    # Pinned to v3: LocalStack 4.x requires LOCALSTACK_AUTH_TOKEN even for community services
+    # (kinesis/glue here) and exits with code 55 on startup otherwise. v3 (last 3.8.1) is the
+    # final auth-free release.
+    image: 'localstack/localstack:3'
     environment:
       - SERVICES=kinesis,glue
       - KINESIS_ERROR_PROBABILITY=0 # This option is interesting, but will likely cause test flakes.

